README.md
# skc

`skc` is a simple tool for finding shared k-mer content between two genomes.

## Installation

### Prebuilt binary

```
curl -sSL skc.mbh.sh | sh
# or with wget
wget -nv -O - skc.mbh.sh | sh
```

You can also pass options to the script like so

```text
$ curl -sSL skc.mbh.sh | sh -s -- --help
install.sh [option]

Fetch and install the latest version of skc, if skc is already
installed it will be updated to the latest version.

Options
        -V, --verbose
                Enable verbose output for the installer

        -f, -y, --force, --yes
                Skip the confirmation prompt during installation

        -p, --platform
                Override the platform identified by the installer

        -b, --bin-dir
                Override the bin installation directory [default: /usr/local/bin]

        -a, --arch
                Override the architecture identified by the installer [default: x86_64]

        -B, --base-url
                Override the base URL used for downloading releases [default: https://github.com/mbhall88/skc/releases]

        -h, --help
                Display this help message

```

### Cargo

```text
cargo install skc
```

### Conda

```text
conda install skc
```

### Local

```text
cargo build --release
./target/release/skc --help
```

## Usage

Check for shared 16-mers between the [HIV-1 genome](https://www.ncbi.nlm.nih.gov/nuccore/NC_001802.1) and the [
*Mycobacterium tuberculosis* genome](https://www.ncbi.nlm.nih.gov/nuccore/NC_000962.3).

```text
$ skc -k 16 NC_001802.1.fa NC_000962.3.fa
[2023-06-20T01:46:36Z INFO ] 9079 unique k-mers in target
[2023-06-20T01:46:38Z INFO ] 2 shared k-mers between target and query
>4233642782 tcount=1 qcount=1 tpos=NC_001802.1:739 qpos=NC_000962.3:4008106
TGCAGAACATCCAGGG
>4237062597 tcount=1 qcount=1 tpos=NC_001802.1:8415 qpos=NC_000962.3:629482
CCAGCAGCAGATAGGG
```

So we can see there are two shared 16-mers between the genomes. By default, the shared k-mers are written to stdout -
use the `-o` option to write them to file.

### Fasta description

Example: `>4233642782 tcount=1 qcount=1 tpos=NC_001802.1:739 qpos=NC_000962.3:4008106`

The ID (`4233642782`) is the 64-bit integer representation of the k-mer's value in bit-space (
see [Daniel Liu's brilliant `cute-nucleotides` repository][cute] for more information). `tcount` and `qcount` are the
number of times the k-mer is present in the target and query genomes, respectively. `tpos` and `qpos` are the (1-based)
k-mer starting position(s) within the target and query contigs - these will be comma-seperated if the k-mer occurs
multiple times.

### Usage help

```text
$ skc --help
Shared k-mer content between two genomes

Usage: skc [OPTIONS] <TARGET> <QUERY>

Arguments:
  <TARGET>
          Target sequence

          Can be compressed with gzip, bzip2, xz, or zstd

  <QUERY>
          Query sequence

          Can be compressed with gzip, bzip2, xz, or zstd

Options:
  -k, --kmer <KMER>
          Size of k-mers (max. 32)

          [default: 21]

  -o, --output <OUTPUT>
          Output filepath(s); stdout if not present

  -O, --output-type <u|b|g|l|z>
          u: uncompressed; b: Bzip2; g: Gzip; l: Lzma; z: Zstd

          Output compression format is automatically guessed from the filename extension. This option is used to override that

          [default: u]

  -l, --compress-level <INT>
          Compression level to use if compressing output

          [default: 6]

  -h, --help
          Print help (see a summary with '-h')

  -V, --version
          Print version
```

### Caveats

- Make the first genome passed (`<TARGET>`) the smallest genome. This is to reduce memory usage as all unique k-mers (
  well their `u64` value) for this genome will be held in memory.
- We do not use canonical k-mers
- 32 is the largest k-mer size that can be used. This is basically a (lazy) implementation decision, but also helps to
  keep the memory footprint as low as possible. If you want larger k-mer values, I would suggest checking out some of
  the [similar tools](#alternate-tools).

## Alternate tools

`skc` does not claim to be the fastest or most memory-efficient tool to find shared k-mer content. I basically wrote it
as I either struggled to install some alternate tools, they were clunky/verbose, or it was laborious to get shared
k-mers out of the results (e.g. can only search one k-mer at a time or have to run many different subcommands). Here is
a (non-exhaustive) list of other tools that can be used to get shared k-mer content

- [unikmer](https://github.com/shenwei356/unikmer) - this was brought to my attention *after* I wrote `skc`. Had I known
  about it beforehand, I probably wouldn't have written `skc`. So I would recommend unikmer for almost all use
  cases - [Wei Shen](https://github.com/shenwei356) writes awesome tools
- [Jellyfish](https://github.com/gmarcais/Jellyfish)
- [REINDEER](https://github.com/kamimrcht/REINDEER)
- [kmer-db](https://github.com/refresh-bio/kmer-db)
- [GGCAT](https://github.com/algbio/ggcat)
- [KAT](https://github.com/TGAC/KAT)

## Acknowledgements

[Daniel Liu's brilliant `cute-nucleotides` repository][cute] is used to (rapidly) convert k-mers into 64-bit integers.

[cute]: https://github.com/Daniel-Liu-c0deb0t/cute-nucleotides



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]


[unreleased]: https://github.com/olivierlacan/keep-a-changelog/compare/0.1.0...HEAD

src/lib.rs
#[cfg(target_arch = "x86")]
use std::arch::x86::*;
#[cfg(target_arch = "x86_64")]
use std::arch::x86_64::*;

use std::alloc;
use std::ffi::OsStr;
use std::path::Path;
use thiserror::Error;

#[derive(Default, Debug, Clone, Eq, PartialEq, Hash)]
pub struct KmerInfo {
    count: u32,
    pub positions: Vec<String>,
}

impl KmerInfo {
    pub fn new() -> Self {
        Self::default()
    }

    /// Add a position to the KmerInfo. `chrom` and `pos` will be joined as `chrom`:`pos`. `pos` is
    /// expected to be zero-based and will be incremented by 1 when joined with `chrom` to make it
    /// one-based, which makes for easy searching in genome browsers
    pub fn add_pos(&mut self, chrom: &str, pos: usize) {
        let s = format!("{chrom}:{}", pos + 1);
        self.positions.push(s)
    }

    pub fn count(&self) -> usize {
        self.positions.len()
    }
}

pub fn encode(nuc: &[u8]) -> Vec<u64> {
    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
    {
        if is_x86_feature_detected!("avx2") {
            return unsafe { encode_movemask_avx(nuc) };
        } else if is_x86_feature_detected!("sse2") {
            return unsafe { encode_movemask_sse(nuc) };
        }
    }

    encode_lut(nuc)
}

#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
#[target_feature(enable = "avx2")]
unsafe fn encode_movemask_avx(nuc: &[u8]) -> Vec<u64> {
    let ptr = nuc.as_ptr() as *const __m256i;
    let end_idx = nuc.len() / 32;
    let len = end_idx + if nuc.len() % 32 == 0 { 0 } else { 1 };

    let layout = alloc::Layout::from_size_align_unchecked(len * 8, 32);
    let res_ptr = alloc::alloc(layout) as *mut u64;

    for i in 0..end_idx as isize {
        let v = _mm256_loadu_si256(ptr.offset(i));

        // permute because unpacks works on the low/high 64 bits in each lane
        let v = _mm256_permute4x64_epi64(v, 0b11011000);

        // shift each group of two bits for each nucleotide to the end of each byte
        let lo = _mm256_slli_epi64(v, 6);
        let hi = _mm256_slli_epi64(v, 5);

        // interleave bytes then extract the bit at the end of each byte
        let a = _mm256_unpackhi_epi8(lo, hi);
        let b = _mm256_unpacklo_epi8(lo, hi);

        // zero extend after movemask
        let a = (_mm256_movemask_epi8(a) as u32) as u64;
        let b = (_mm256_movemask_epi8(b) as u32) as u64;

        *res_ptr.offset(i) = (a << 32) | b;
    }

    if nuc.len() % 32 > 0 {
        *res_ptr.add(end_idx) = *encode_lut(&nuc[(end_idx * 32)..]).get_unchecked(0);
    }

    Vec::from_raw_parts(res_ptr, len, len)
}

#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
#[target_feature(enable = "sse2")]
unsafe fn encode_movemask_sse(nuc: &[u8]) -> Vec<u64> {
    let ptr = nuc.as_ptr() as *const __m128i;
    let end_idx = nuc.len() / 16;
    let len = nuc.len() / 32 + if nuc.len() % 32 == 0 { 0 } else { 1 };

    let layout = alloc::Layout::from_size_align_unchecked(len * 8, 16);
    let res_ptr = alloc::alloc(layout) as *mut u32;

    for i in 0..end_idx as isize {
        let v = _mm_loadu_si128(ptr.offset(i));

        // shift each group of two bits for each nucleotide to the end of each byte
        let lo = _mm_slli_epi64(v, 6);
        let hi = _mm_slli_epi64(v, 5);

        // interleave bytes then extract the bit at the end of each byte
        let a = _mm_unpackhi_epi8(lo, hi);
        let b = _mm_unpacklo_epi8(lo, hi);
        let a = _mm_movemask_epi8(a);
        let b = _mm_movemask_epi8(b);

        *res_ptr.offset(i) = ((a << 16) | b) as u32;
    }

    if nuc.len() % 16 > 0 {
        *res_ptr.add(end_idx) = *encode_lut(&nuc[(end_idx * 16)..]).get_unchecked(0) as u32;
    }

    Vec::from_raw_parts(res_ptr as *mut u64, len, len)
}

static BYTE_LUT: [u8; 128] = {
    let mut lut = [0u8; 128];
    lut[b'a' as usize] = 0b00;
    lut[b't' as usize] = 0b10;
    lut[b'u' as usize] = 0b10;
    lut[b'c' as usize] = 0b01;
    lut[b'g' as usize] = 0b11;
    lut[b'A' as usize] = 0b00;
    lut[b'T' as usize] = 0b10;
    lut[b'U' as usize] = 0b10;
    lut[b'C' as usize] = 0b01;
    lut[b'G' as usize] = 0b11;
    lut
};

fn encode_lut(nuc: &[u8]) -> Vec<u64> {
    let mut res = vec![0u64; (nuc.len() / 32) + if nuc.len() % 32 == 0 { 0 } else { 1 }];

    for i in 0..nuc.len() {
        let offset = i / 32;
        let shift = (i % 32) << 1;

        unsafe {
            *res.get_unchecked_mut(offset) = *res.get_unchecked(offset)
                | ((*BYTE_LUT.get_unchecked(*nuc.get_unchecked(i) as usize) as u64) << shift);
        }
    }

    res
}

pub fn decode(bits: &[u64], len: usize) -> Vec<u8> {
    if len > (bits.len() * 32) {
        panic!(
            "The length {} is greater than the number of nucleotides!",
            len
        );
    }

    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
    {
        if is_x86_feature_detected!("avx2") {
            return unsafe { decode_shuffle_avx(bits, len) };
        } else if is_x86_feature_detected!("sse4.1") {
            return unsafe { decode_shuffle_sse(bits, len) };
        }
    }

    decode_lut(bits, len)
}

#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
#[target_feature(enable = "avx2")]
unsafe fn decode_shuffle_avx(bits: &[u64], len: usize) -> Vec<u8> {
    let layout = alloc::Layout::from_size_align_unchecked(bits.len() * 32, 32);
    let ptr = alloc::alloc(layout) as *mut __m256i;

    let shuffle_mask = _mm256_set_epi32(
        0x07070707, 0x06060606, 0x05050505, 0x04040404, 0x03030303, 0x02020202, 0x01010101,
        0x00000000,
    );
    let lo_mask = _mm256_set1_epi16(0b0000110000000011);
    let lut_i32 =
        (b'A' as i32) | ((b'C' as i32) << 8) | ((b'T' as i32) << 16) | ((b'G' as i32) << 24);
    let lut = _mm256_set_epi32(
        b'G' as i32,
        b'T' as i32,
        b'C' as i32,
        lut_i32,
        b'G' as i32,
        b'T' as i32,
        b'C' as i32,
        lut_i32,
    );

    for i in 0..bits.len() {
        let curr = *bits.get_unchecked(i) as i64;
        let v = _mm256_set1_epi64x(curr);

        // duplicate each byte four times
        let v1 = _mm256_shuffle_epi8(v, shuffle_mask);

        // separately right shift each 16-bit chunk by 0 or 4 bits
        let v2 = _mm256_srli_epi16(v1, 4);

        // merge together shifted chunks
        let v = _mm256_blend_epi16(v1, v2, 0b10101010i32);

        // only keep two bits in each byte
        // either 0b0011 or 0b1100
        let v = _mm256_and_si256(v, lo_mask);

        // use lookup table to convert nucleotide bits to bytes
        let v = _mm256_shuffle_epi8(lut, v);
        _mm256_store_si256(ptr.add(i), v);
    }

    Vec::from_raw_parts(ptr as *mut u8, len, bits.len() * 32)
}

#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
#[target_feature(enable = "sse4.1")]
unsafe fn decode_shuffle_sse(bits: &[u64], len: usize) -> Vec<u8> {
    let layout = alloc::Layout::from_size_align_unchecked(bits.len() * 32, 16);
    let ptr = alloc::alloc(layout) as *mut __m128i;

    let bits_ptr = bits.as_ptr() as *const i32;

    let shuffle_mask = _mm_set_epi32(0x03030303, 0x02020202, 0x01010101, 0x00000000);
    let lo_mask = _mm_set1_epi16(0b0000110000000011);
    let lut_i32 =
        (b'A' as i32) | ((b'C' as i32) << 8) | ((b'T' as i32) << 16) | ((b'G' as i32) << 24);
    let lut = _mm_set_epi32(b'G' as i32, b'T' as i32, b'C' as i32, lut_i32);

    for i in 0..(bits.len() * 2) as isize {
        let curr = *bits_ptr.offset(i);
        let v = _mm_set1_epi32(curr);

        // duplicate each byte four times
        let v1 = _mm_shuffle_epi8(v, shuffle_mask);

        // separately right shift each 16-bit chunk by 0 or 4 bits
        let v2 = _mm_srli_epi16(v1, 4);

        // merge together shifted chunks
        let v = _mm_blend_epi16(v1, v2, 0b10101010i32);

        // only keep two bits in each byte
        // either 0b0011 or 0b1100
        let v = _mm_and_si128(v, lo_mask);

        // use lookup table to convert nucleotide bits to bytes
        let v = _mm_shuffle_epi8(lut, v);
        _mm_store_si128(ptr.offset(i), v);
    }

    Vec::from_raw_parts(ptr as *mut u8, len, bits.len() * 32)
}

static BITS_LUT: [u8; 4] = {
    let mut lut = [0u8; 4];
    lut[0b00] = b'A';
    lut[0b10] = b'T';
    lut[0b01] = b'C';
    lut[0b11] = b'G';
    lut
};

fn decode_lut(bits: &[u64], len: usize) -> Vec<u8> {
    let layout = unsafe { alloc::Layout::from_size_align_unchecked(len, 1) };
    let res_ptr = unsafe { alloc::alloc(layout) };

    for i in 0..len {
        let offset = i >> 5;
        let shift = (i & 31) << 1;
        let curr = unsafe { *bits.get_unchecked(offset) };

        unsafe {
            *res_ptr.add(i) = *BITS_LUT.get_unchecked(((curr >> shift) & 0b11) as usize);
        }
    }

    unsafe { Vec::from_raw_parts(res_ptr, len, len) }
}

/// A collection of custom errors relating to the command line interface for this package.
#[derive(Error, Debug, PartialEq)]
pub enum CliError {
    /// Indicates that a string cannot be parsed into a [`CompressionFormat`](#compressionformat).
    #[error("{0} is not a valid output format")]
    InvalidCompression(String),
}

pub trait CompressionExt {
    fn from_path<S: AsRef<OsStr> + ?Sized>(p: &S) -> Self;
}

impl CompressionExt for niffler::compression::Format {
    /// Attempts to infer the compression type from the file extension. If the extension is not
    /// known, then Uncompressed is returned.
    fn from_path<S: AsRef<OsStr> + ?Sized>(p: &S) -> Self {
        let path = Path::new(p);
        match path.extension().map(|s| s.to_str()) {
            Some(Some("gz")) => Self::Gzip,
            Some(Some("zst")) => Self::Zstd,
            Some(Some("bz") | Some("bz2")) => Self::Bzip,
            Some(Some("lzma")) => Self::Lzma,
            _ => Self::No,
        }
    }
}

pub fn parse_compression_format(s: &str) -> Result<niffler::compression::Format, CliError> {
    match s {
        "b" | "B" => Ok(niffler::Format::Bzip),
        "g" | "G" => Ok(niffler::Format::Gzip),
        "l" | "L" => Ok(niffler::Format::Lzma),
        "u" | "U" => Ok(niffler::Format::No),
        "z" | "Z" => Ok(niffler::Format::Zstd),
        _ => Err(CliError::InvalidCompression(s.to_string())),
    }
}

/// A utility function to validate compression level is in allowed range
#[allow(clippy::redundant_clone)]
pub fn parse_level(s: &str) -> Result<niffler::Level, String> {
    let lvl = match s.parse::<u8>() {
        Ok(1) => niffler::Level::One,
        Ok(2) => niffler::Level::Two,
        Ok(3) => niffler::Level::Three,
        Ok(4) => niffler::Level::Four,
        Ok(5) => niffler::Level::Five,
        Ok(6) => niffler::Level::Six,
        Ok(7) => niffler::Level::Seven,
        Ok(8) => niffler::Level::Eight,
        Ok(9) => niffler::Level::Nine,
        Ok(10) => niffler::Level::Ten,
        Ok(11) => niffler::Level::Eleven,
        Ok(12) => niffler::Level::Twelve,
        Ok(13) => niffler::Level::Thirteen,
        Ok(14) => niffler::Level::Fourteen,
        Ok(15) => niffler::Level::Fifteen,
        Ok(16) => niffler::Level::Sixteen,
        Ok(17) => niffler::Level::Seventeen,
        Ok(18) => niffler::Level::Eighteen,
        Ok(19) => niffler::Level::Nineteen,
        Ok(20) => niffler::Level::Twenty,
        Ok(21) => niffler::Level::TwentyOne,
        _ => return Err(format!("Compression level {} not in the range 1-21", s)),
    };
    Ok(lvl)
}

#[cfg(test)]
mod tests {
    use super::*;
    // A = 00, T/U = 10, C = 01, G = 11
    #[test]
    fn test_encode() {
        assert_eq!(
            encode(b"ATCGATCGATCGATCGATCGATCGATCGATCG"),
            vec![0b1101100011011000110110001101100011011000110110001101100011011000]
        );
        assert_eq!(encode(b"ATCG"), vec![0b11011000]);
        assert_eq!(encode(b"atcg"), vec![0b11011000]);
    }

    #[test]
    fn test_decode() {
        assert_eq!(
            decode(
                &[0b1101100011011000110110001101100011011000110110001101100011011000],
                32
            ),
            "ATCGATCGATCGATCGATCGATCGATCGATCG".as_bytes()
        );
    }

    #[test]
    fn test_add_pos() {
        let mut ki = KmerInfo::new();
        ki.add_pos("chr", 3);
        assert_eq!(ki.count(), 1);

        ki.add_pos("chr", 9);
        assert_eq!(ki.count(), 2);
        assert_eq!(ki.positions, vec!["chr:4", "chr:10"]);
    }
}


install/install.sh
#!/usr/bin/env sh

set -eu
printf '\n'

BOLD="$(tput bold 2>/dev/null || printf '')"
GREY="$(tput setaf 0 2>/dev/null || printf '')"
UNDERLINE="$(tput smul 2>/dev/null || printf '')"
RED="$(tput setaf 1 2>/dev/null || printf '')"
GREEN="$(tput setaf 2 2>/dev/null || printf '')"
YELLOW="$(tput setaf 3 2>/dev/null || printf '')"
BLUE="$(tput setaf 4 2>/dev/null || printf '')"
MAGENTA="$(tput setaf 5 2>/dev/null || printf '')"
NO_COLOR="$(tput sgr0 2>/dev/null || printf '')"
PROJECT="skc"
GH_USER="mbhall88"

SUPPORTED_TARGETS="x86_64-unknown-linux-musl x86_64-unknown-linux-gnu \
                   aarch64-unknown-linux-musl x86_64-apple-darwin"

info() {
  printf '%s\n' "${BOLD}${GREY}>${NO_COLOR} $*"
}

warn() {
  printf '%s\n' "${YELLOW}! $*${NO_COLOR}"
}

error() {
  printf '%s\n' "${RED}x $*${NO_COLOR}" >&2
}

completed() {
  printf '%s\n' "${GREEN}✓${NO_COLOR} $*"
}

has() {
  command -v "$1" 1>/dev/null 2>&1
}

# Make sure user is not using zsh or non-POSIX-mode bash, which can cause issues
verify_shell_is_posix_or_exit() {
  if [ -n "${ZSH_VERSION+x}" ]; then
    error "Running installation script with \`zsh\` is known to cause errors."
    error "Please use \`sh\` instead."
    exit 1
  elif [ -n "${BASH_VERSION+x}" ] && [ -z "${POSIXLY_CORRECT+x}" ]; then
    error "Running installation script with non-POSIX \`bash\` may cause errors."
    error "Please use \`sh\` instead."
    exit 1
  else
    true # No-op: no issues detected
  fi
}

# Gets path to a temporary file, even if
get_tmpfile() {
  suffix="$1"
  if has mktemp; then
    printf "%s.%s" "$(mktemp)" "${suffix}"
  else
    # No really good options here--let's pick a default + hope
    printf "/tmp/${PROJECT}.%s" "${suffix}"
  fi
}

# Makes a temporary directory
get_tmpdir() {
  if has mktemp; then
    printf "%s" "$(mktemp -d)"
  else
    # No really good options here--let's pick a default + hope
    printf "%s" "/tmp/${PROJECT}"
  fi
}

# Gets the latest github release tag (version)
get_version_from_github() {
  ver=$(wget -qO - "https://api.github.com/repos/${GH_USER}/${PROJECT}/releases/latest" |
    grep '"tag_name":' | sed -E 's/.*"([^"]+)".*/\1/')
  echo "$ver"
}

# Test if a location is writeable by trying to write to it. Windows does not let
# you test writeability other than by writing: https://stackoverflow.com/q/1999988
test_writeable() {
  path="${1:-}/test.txt"
  if touch "${path}" 2>/dev/null; then
    rm "${path}"
    return 0
  else
    return 1
  fi
}

download() {
  file="$1"
  url="$2"

  if has curl; then
    cmd="curl --fail --silent --location --output $file $url"
  elif has wget; then
    cmd="wget --quiet --output-document=$file $url"
  elif has fetch; then
    cmd="fetch --quiet --output=$file $url"
  else
    error "No HTTP download program (curl, wget, fetch) found, exiting…"
    return 1
  fi

  $cmd && return 0 || rc=$?

  error "Command failed (exit code $rc): ${BLUE}${cmd}${NO_COLOR}"
  printf "\n" >&2
  info "This is likely due to $PROJECT not yet supporting your configuration."
  info "If you would like to see a build for your configuration,"
  info "please create an issue requesting a build for ${MAGENTA}${TARGET}${NO_COLOR}:"
  info "${BOLD}${UNDERLINE}https://github.com/${GH_USER}/${PROJECT}/issues/new/${NO_COLOR}"
  return $rc
}

unpack() {
  archive=$1
  bin_dir=$2
  temp_dir=$(get_tmpdir)
  sudo=${3-}

  case "$archive" in
  *.tar.gz)
    flags=$(test -n "${VERBOSE-}" && echo "-xzvof" || echo "-xzof")
    ${sudo} tar "${flags}" "${archive}" -C "${temp_dir}"
    ${sudo} mv "${temp_dir}"/${PROJECT}-*/${PROJECT} "${bin_dir}/${PROJECT}"
    return 0
    ;;
  *.zip)
    flags=$(test -z "${VERBOSE-}" && echo "-qqo" || echo "-o")
    UNZIP="${flags}" ${sudo} unzip "${archive}" -d "${temp_dir}"
    return 0
    ;;
  esac

  error "Unknown package extension."
  printf "\n"
  info "This almost certainly results from a bug in this script--please file a"
  info "bug report at https://github.com/${GH_USER}/${PROJECT}/issues"
  return 1
}

usage() {
  printf "%s\n" \
    "install.sh [option]" \
    "" \
    "Fetch and install the latest version of $PROJECT, if $PROJECT is already" \
    "installed it will be updated to the latest version."

  printf "\n%s\n" "Options"
  printf "\t%s\n\t\t%s\n\n" \
    "-V, --verbose" "Enable verbose output for the installer" \
    "-f, -y, --force, --yes" "Skip the confirmation prompt during installation" \
    "-p, --platform" "Override the platform identified by the installer [default: ${PLATFORM}]" \
    "-b, --bin-dir" "Override the bin installation directory [default: ${BIN_DIR}]" \
    "-a, --arch" "Override the architecture identified by the installer [default: ${ARCH}]" \
    "-B, --base-url" "Override the base URL used for downloading releases [default: ${BASE_URL}]" \
    "-h, --help" "Display this help message"
}

elevate_priv() {
  if ! has sudo; then
    error 'Could not find the command "sudo", needed to get permissions for install.'
    info "If you are on Windows, please run your shell as an administrator, then"
    info "rerun this script. Otherwise, please run this script as root, or install"
    info "sudo."
    exit 1
  fi
  if ! sudo -v; then
    error "Superuser not granted, aborting installation"
    exit 1
  fi
}

install() {
  ext="$1"
  url="$2"

  if test_writeable "${BIN_DIR}"; then
    sudo=""
    msg="Installing $PROJECT, please wait…"
  else
    warn "Escalated permissions are required to install to ${BIN_DIR}"
    elevate_priv
    sudo="sudo"
    msg="Installing $PROJECT as root, please wait…"
  fi
  info "$msg"

  archive=$(get_tmpfile "$ext")

  # download to the temp file
  download "${archive}" "${url}"

  # unpack the temp file to the bin dir, using sudo if required
  unpack "${archive}" "${BIN_DIR}" "${sudo}"
}

# Currently supporting:
#   - win (Git Bash)
#   - darwin
#   - linux
#   - linux_musl (Alpine)
#   - freebsd
detect_platform() {
  platform="$(uname -s | tr '[:upper:]' '[:lower:]')"

  case "${platform}" in
  msys_nt*) platform="pc-windows-msvc" ;;
  cygwin_nt*) platform="pc-windows-msvc" ;;
  # mingw is Git-Bash
  mingw*) platform="pc-windows-msvc" ;;
  # use the statically compiled musl bins on linux to avoid linking issues.
  linux) platform="unknown-linux-musl" ;;
  darwin) platform="apple-darwin" ;;
  freebsd) platform="unknown-freebsd" ;;
  esac

  printf '%s' "${platform}"
}

# Currently supporting:
#   - x86_64
#   - i386
#   - arm
#   - arm64
detect_arch() {
  arch="$(uname -m | tr '[:upper:]' '[:lower:]')"

  case "${arch}" in
  amd64) arch="x86_64" ;;
  armv*) arch="arm" ;;
  arm64) arch="aarch64" ;;
  esac

  # `uname -m` in some cases mis-reports 32-bit OS as 64-bit, so double check
  if [ "${arch}" = "x86_64" ] && [ "$(getconf LONG_BIT)" -eq 32 ]; then
    arch=i686
  elif [ "${arch}" = "aarch64" ] && [ "$(getconf LONG_BIT)" -eq 32 ]; then
    arch=arm
  fi

  printf '%s' "${arch}"
}

detect_target() {
  arch="$1"
  platform="$2"
  target="$arch-$platform"

  if [ "${target}" = "arm-unknown-linux-musl" ]; then
    target="${target}eabihf"
  fi

  printf '%s' "${target}"
}

confirm() {
  if [ -z "${FORCE-}" ]; then
    printf "%s " "${MAGENTA}?${NO_COLOR} $* ${BOLD}[y/N]${NO_COLOR}"
    set +e
    read -r yn </dev/tty
    rc=$?
    set -e
    if [ $rc -ne 0 ]; then
      error "Error reading from prompt (please re-run with the '--yes' option)"
      exit 1
    fi
    if [ "$yn" != "y" ] && [ "$yn" != "yes" ]; then
      error 'Aborting (please answer "yes" to continue)'
      exit 1
    fi
  fi
}

check_bin_dir() {
  bin_dir="${1%/}"

  if [ ! -d "$BIN_DIR" ]; then
    error "Installation location $BIN_DIR does not appear to be a directory"
    info "Make sure the location exists and is a directory, then try again."
    usage
    exit 1
  fi

  # https://stackoverflow.com/a/11655875
  good=$(
    IFS=:
    for path in $PATH; do
      if [ "${path%/}" = "${bin_dir}" ]; then
        printf 1
        break
      fi
    done
  )

  if [ "${good}" != "1" ]; then
    warn "Bin directory ${bin_dir} is not in your \$PATH"
  fi
}

is_build_available() {
  arch="$1"
  platform="$2"
  target="$3"

  good=$(
    IFS=" "
    for t in $SUPPORTED_TARGETS; do
      if [ "${t}" = "${target}" ]; then
        printf 1
        break
      fi
    done
  )

  if [ "${good}" != "1" ]; then
    error "${arch} builds for ${platform} are not yet available for $PROJECT"
    printf "\n" >&2
    info "If you would like to see a build for your configuration,"
    info "please create an issue requesting a build for ${MAGENTA}${target}${NO_COLOR}:"
    info "${BOLD}${UNDERLINE}https://github.com/${GH_USER}/${PROJECT}/issues/new/${NO_COLOR}"
    printf "\n"
    exit 1
  fi
}

# defaults
if [ -z "${PLATFORM-}" ]; then
  PLATFORM="$(detect_platform)"
fi

if [ -z "${BIN_DIR-}" ]; then
  BIN_DIR=/usr/local/bin
fi

if [ -z "${ARCH-}" ]; then
  ARCH="$(detect_arch)"
fi

if [ -z "${BASE_URL-}" ]; then
  BASE_URL="https://github.com/${GH_USER}/${PROJECT}/releases"
fi

# Non-POSIX shells can break once executing code due to semantic differences
verify_shell_is_posix_or_exit

# parse argv variables
while [ "$#" -gt 0 ]; do
  case "$1" in
  -p | --platform)
    PLATFORM="$2"
    shift 2
    ;;
  -b | --bin-dir)
    BIN_DIR="$2"
    shift 2
    ;;
  -a | --arch)
    ARCH="$2"
    shift 2
    ;;
  -B | --base-url)
    BASE_URL="$2"
    shift 2
    ;;

  -V | --verbose)
    VERBOSE=1
    shift 1
    ;;
  -f | -y | --force | --yes)
    FORCE=1
    shift 1
    ;;
  -h | --help)
    usage
    exit
    ;;

  -p=* | --platform=*)
    PLATFORM="${1#*=}"
    shift 1
    ;;
  -b=* | --bin-dir=*)
    BIN_DIR="${1#*=}"
    shift 1
    ;;
  -a=* | --arch=*)
    ARCH="${1#*=}"
    shift 1
    ;;
  -B=* | --base-url=*)
    BASE_URL="${1#*=}"
    shift 1
    ;;
  -V=* | --verbose=*)
    VERBOSE="${1#*=}"
    shift 1
    ;;
  -f=* | -y=* | --force=* | --yes=*)
    FORCE="${1#*=}"
    shift 1
    ;;

  *)
    error "Unknown option: $1"
    usage
    exit 1
    ;;
  esac
done

TARGET="$(detect_target "${ARCH}" "${PLATFORM}")"

is_build_available "${ARCH}" "${PLATFORM}" "${TARGET}"

printf "  %s\n" "${UNDERLINE}Configuration${NO_COLOR}"
info "${BOLD}Bin directory${NO_COLOR}: ${GREEN}${BIN_DIR}${NO_COLOR}"
info "${BOLD}Platform${NO_COLOR}:      ${GREEN}${PLATFORM}${NO_COLOR}"
info "${BOLD}Arch${NO_COLOR}:          ${GREEN}${ARCH}${NO_COLOR}"

# non-empty VERBOSE enables verbose untarring
if [ -n "${VERBOSE-}" ]; then
  VERBOSE=v
  info "${BOLD}Verbose${NO_COLOR}: yes"
else
  VERBOSE=
fi

printf '\n'

EXT=tar.gz
if [ "${PLATFORM}" = "pc-windows-msvc" ]; then
  EXT=zip
fi

VERSION=$(get_version_from_github)
URL="${BASE_URL}/latest/download/${PROJECT}-${VERSION}-${TARGET}.${EXT}"
info "Tarball URL: ${UNDERLINE}${BLUE}${URL}${NO_COLOR}"
confirm "Install $PROJECT ${GREEN}v${VERSION} (latest)${NO_COLOR} to ${BOLD}${GREEN}${BIN_DIR}${NO_COLOR}?"
check_bin_dir "${BIN_DIR}"

install "${EXT}" "$URL"
completed "$PROJECT (v${VERSION}) installed"
